packages/backend/src/app/workers/flow-worker/flow-worker.ts

```diff
@@ -70,20 +71,30 @@ async function downloadFiles(
     flowVersion: FlowVersion,
     collectionVersion: CollectionVersion,
 ): Promise<void> {
-    const buildPath = sandbox.getSandboxFolderPath();
+    const flowLock = await createRedisLock();
+    try {
+        logger.info(`[${flowVersion.id}] Acquiring flow lock to build codes`);
+        await flowLock.acquire(flowVersion.id);
+
+        const buildPath = sandbox.getSandboxFolderPath();
 
-    // This has to be before flows, since it does modify code settings and fill it with packaged file id.
-    fs.mkdirSync(buildPath + "/codes/");
-    const artifacts: File[] = await buildCodes(projectId, flowVersion);
-    artifacts.forEach((artifact) => {
-        fs.writeFileSync(buildPath + "/codes/" + artifact.id + ".js", artifact.data);
-    });
+        // This has to be before flows, since it does modify code settings and fill it with packaged file id.
+        fs.mkdirSync(buildPath + "/codes/");
+        const artifacts: File[] = await buildCodes(projectId, flowVersion);
+        artifacts.forEach((artifact) => {
+            fs.writeFileSync(buildPath + "/codes/" + artifact.id + ".js", artifact.data);
+        });
 
-    fs.mkdirSync(buildPath + "/flows/");
-    fs.writeFileSync(buildPath + "/flows/" + flowVersion.id + ".json", JSON.stringify(flowVersion));
+        fs.mkdirSync(buildPath + "/flows/");
+        fs.writeFileSync(buildPath + "/flows/" + flowVersion.id + ".json", JSON.stringify(flowVersion));
 
-    fs.mkdirSync(buildPath + "/collections/");
-    fs.writeFileSync(buildPath + "/collections/" + collectionVersion.id + ".json", JSON.stringify(collectionVersion));
+        fs.mkdirSync(buildPath + "/collections/");
+        fs.writeFileSync(buildPath + "/collections/" + collectionVersion.id + ".json", JSON.stringify(collectionVersion));
```

Comment:
i think using async fs methods is a higher priority than flow lock, sync methods block the entire service process.